timemachine/fe/single_topology_v3.py

```diff
@@ -483,6 +484,211 @@ def find_dummy_groups_and_anchors(mol_a, mol_b, core_a, core_b):
+    """
+    Interpolate between force constants using a log-linear functional form.
+
+    In the special case when src_k=0 or dst_k=0 (e.g. ring opening or closing transformations):
+
+    1. Intermediates are interpolated from k_min instead of zero (since 0 is not in the range of the interpolation
+       function)
+    2. Interpolation is restricted to the interval [lambda_min, lambda_max] and pinned to the end state values outside
+       of this range
+
+    Parameters
+    ----------
+    src_k, dst_k : float, k >= 0
+        force constants at lambda=0 and lambda=1, respectively
+
+    k_min : float, k_min > 0
+        minimum force constant for interpolation
+
+    lambda_min, lambda_max : float, in 0 < lambda_min < lambda_max < 1
+        values of the control parameter at which to begin and end interpolation, respectively. This is only relevant
+        when k=0 in the src or dst params. Note that if k=0 in the dst params (i.e. ring opening), the convention is
+        "flipped", i.e. interpolation "begins" at lambda = 1 - lambda_min and "ends" at 1 - lambda_max.
+    """
+    return jnp.where(
+        lamb == 0.0,
+        src_k,
+        jnp.where(
+            lamb == 1.0,
+            dst_k,
+            handle_ring_opening_closing(
+                partial(interpolate.log_linear_interpolation, min_value=k_min),
+                src_k,
+                dst_k,
+                lamb,
+                lambda_min,
+                lambda_max,
+            ),
+        ),
+    )
+
+
+def interpolate_harmonic_bond_params(src_params, dst_params, lamb, k_min, lambda_min, lambda_max):
+    """
+    Interpolate harmonic bond parameters using
+
+    1. Log-linear interpolation for force constants*
+    2. Linear interpolation for equilibrium bond lengths
+
+    * see note on special case when src_k=0 or dst_k=0 in the docstring of `interpolate_harmonic_force_constant`.
+
+    Parameters
+    ----------
+    src_params : array-like, float, (2,)
+        force constant and equilibrium length at lambda=0
+
+    dst_params : array-like, float, (2,)
+        force constant and equilibrium length at lambda=1
+
+    lamb : float
+        control parameter
+
+    k_min, lambda_min, lambda_max : float
+        see docstring of `interpolate_harmonic_force_constant` for documentation of these parameters
+    """
+
+    src_k, src_x = src_params
+    dst_k, dst_x = dst_params
+
+    k = interpolate_harmonic_force_constant(src_k, dst_k, lamb, k_min, lambda_min, lambda_max)
+    x = interpolate.linear_interpolation(src_x, dst_x, lamb)
+
+    return jnp.array([k, x])
+
+
+def cyclic_difference(a, b, period):
+    """
+    Returns the minimum difference between two points, with periodic boundaries.
+    I.e. the solution of ::
+
+        (a + d) % period = b % period
+
+    with minimum abs(d).
+    """
+
+    d = jnp.fmod(b - a, period)
+
+    def f(d):
+        return jnp.where(d <= period / 2, d, d - period)
+
+    return jnp.where(0 < d, f(d), -f(-d))
```

Comment:
It feels like there might be a simpler formula / builtin function for this definition of cyclic difference, but I wasn't able to figure it out yet.